src/Z3/Lock.hs

```diff
@@ -0,0 +1,75 @@
+-- |
+-- Module    : Z3.Lock
+-- Copyright : (c) Iago Abal, 2013-2015
+--             (c) David Castro, 2013
+-- License   : BSD3
+-- Maintainer: Iago Abal <mail@iagoabal.eu>,
+--             David Castro <david.castro.dcp@gmail.com>
+--
+-- A minimal implementation of a re-entrant lock, adopted from https://github.com/basvandijk/concurrent-extra
+--
+--
+
+module Z3.Lock where
+
+import Control.Concurrent( MVar, takeMVar, newMVar, tryPutMVar, readMVar )
+import Control.Monad( when )
+
+-- | A lock is in one of two states: \"locked\" or \"unlocked\".
+newtype Lock = Lock {un :: MVar ()} deriving Eq
+
+-- | Create a lock in the \"unlocked\" state.
+new :: IO Lock
+new = Lock <$> newMVar ()
+
+{-|
+Acquires the 'Lock'. Blocks if another thread has acquired the 'Lock'.
+
+@acquire@ behaves as follows:
+
+* When the state is \"unlocked\" @acquire@ changes the state to \"locked\".
+
+* When the state is \"locked\" @acquire@ /blocks/ until a call to 'release' in
+another thread wakes the calling thread. Upon awakening it will change the state
+to \"locked\".
+
+There are two further important properties of @acquire@:
+
+* @acquire@ is single-wakeup. That is, if there are multiple threads blocked on
+@acquire@ and the lock is released, only one thread will be woken up. The
+runtime guarantees that the woken thread completes its @acquire@ operation.
+
+* When multiple threads are blocked on @acquire@, they are woken up in FIFO
+order. This is useful for providing fairness properties of abstractions built
+using locks. (Note that this differs from the Python implementation where the
+wake-up order is undefined.)
+-}
+acquire :: Lock -> IO ()
+acquire = takeMVar . un
+
+{-|
+@release@ changes the state to \"unlocked\" and returns immediately.
+
+Note that it is an error to release a lock in the \"unlocked\" state!
+
+If there are any threads blocked on 'acquire' the thread that first called
+@acquire@ will be woken up.
+-}
+release :: Lock -> IO ()
+release (Lock mv) = do
+  b <- tryPutMVar mv ()
+  when (not b) $ error "Control.Concurrent.Lock.release: Can't release unlocked Lock!"
```

Comment:
Perhaps fix the error message, `s/Control.Concurrent.Lock.release/Z3.Lock.release`.